soci/soci_convert_test.go

```diff
@@ -0,0 +1,172 @@
+	"context"
+	"testing"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+var (
+	linuxX86 = &ocispec.Platform{
+		Architecture: "amd64",
+		OS:           "linux",
+	}
+
+	imageDesc = ocispec.Descriptor{
+		MediaType: ocispec.MediaTypeImageManifest,
+		Digest:    "sha256:1234",
+		Size:      1234,
+		Platform:  linuxX86,
+	}
+
+	sociIndexDesc = ocispec.Descriptor{
+		MediaType:    ocispec.MediaTypeImageManifest,
+		ArtifactType: SociIndexArtifactTypeV2,
+		Digest:       "sha256:5678",
+		Size:         5678,
+	}
+
+	sociIndexDesc2 = ocispec.Descriptor{
+		MediaType:    ocispec.MediaTypeImageManifest,
+		ArtifactType: SociIndexArtifactTypeV2,
+		Digest:       "sha256:9999",
+		Size:         9999,
+	}
+
+	sociIndexDescWithPlatform  = sociIndexDesc
+	sociIndexDesc2WithPlatform = sociIndexDesc2
+)
+
+func init() {
+	sociIndexDescWithPlatform.Platform = linuxX86
+	sociIndexDesc2WithPlatform.Platform = linuxX86
+}
+
+func TestAddSociIndexes(t *testing.T) {
+	tests := []struct {
+		name        string
+		ociIndex    ocispec.Index
+		sociIndexes []IndexWithMetadata
+		expected    ocispec.Index
+	}{
+		{
+			name: "add soci index",
+			ociIndex: ocispec.Index{
+				MediaType: ocispec.MediaTypeImageIndex,
+				Manifests: []ocispec.Descriptor{imageDesc},
+			},
+			sociIndexes: []IndexWithMetadata{
+				{
+					Desc:     sociIndexDesc,
+					Platform: linuxX86,
+				},
+			},
+			expected: ocispec.Index{
+				MediaType: ocispec.MediaTypeImageIndex,
+				Manifests: []ocispec.Descriptor{
+					imageDesc,
+					sociIndexDescWithPlatform,
+				},
+			},
+		},
+		{
+			name: "add existing soci index to an oci index",
+			ociIndex: ocispec.Index{
+				MediaType: ocispec.MediaTypeImageIndex,
+				Manifests: []ocispec.Descriptor{
+					imageDesc,
+					sociIndexDescWithPlatform,
+				},
+			},
+			sociIndexes: []IndexWithMetadata{
+				{
+					Desc:     sociIndexDesc,
+					Platform: linuxX86,
+				},
+			},
+			expected: ocispec.Index{
+				MediaType: ocispec.MediaTypeImageIndex,
+				Manifests: []ocispec.Descriptor{
+					imageDesc,
+					sociIndexDescWithPlatform,
+				},
+			},
+		},
+		{
+			name: "replace existing soci index in an oci index",
+			ociIndex: ocispec.Index{
+				MediaType: ocispec.MediaTypeImageIndex,
+				Manifests: []ocispec.Descriptor{
+					imageDesc,
+					sociIndexDescWithPlatform,
+				},
+			},
+			sociIndexes: []IndexWithMetadata{
+				{
+					Desc:     sociIndexDesc2,
+					Platform: linuxX86,
+				},
+			},
+			expected: ocispec.Index{
+				MediaType: ocispec.MediaTypeImageIndex,
+				Manifests: []ocispec.Descriptor{
+					imageDesc,
+					sociIndexDesc2WithPlatform,
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			builder := IndexBuilder{}
+			actual := test.ociIndex
+			var indexes []*IndexWithMetadata
+			for _, index := range test.sociIndexes {
+				indexes = append(indexes, &index)
+			}
+			err := builder.addSociIndexes(context.Background(), &actual, indexes)
```

Comment:
Are you planning to add more test? Specially `annotateImages` method, which seems to be handling a lot of image conversion logic.

Reply:
Not in this PR. All of the other methods have side effects like writing new objects into the blobStore or reading from the content store. I'd rather refactor this to isolate those dependencies rather than mocking them out, but I think that would be a lot for this PR. 

For now, we're validating the rest of the logic via integration tests.